### PR TITLE
Move `cudf::test::make_counting_transform_iterator` to `cudf/detail/iterator.cuh`

### PR DESCRIPTION
**Pull request by codereport**
:sob: :sob: :sob: 

Oh how fragile cuDF is. Ran into so many issues. 

This was split out of https://github.com/rapidsai/cudf/pull/6546 because of how complicated it was.

#### 95% was just:
* find & replace
* adding headers

#### The other 5% was:
* issues with `column_device_view.cuh`
* missing dependent headers

### CODE REVIEW

**Review by trxcllnt — APPROVED** [review]
Assuming this passes CI, it looks good to me.

Clang has always complained about the conditional `#ifdef __CUDACC__` blocks in `bit.hpp`, but I assume fixing that is out of scope for this PR.

File: cpp/include/cudf/column/column_device_view.cuh [hunk 1]
```diff
@@ -592,6 +593,7 @@ class alignas(16) mutable_column_device_view : public detail::column_device_view
     return d_children[child_index];
   }
 
+#ifdef __CUDACC__  // because set_bit in bit.hpp is wrapped with __CUDACC__
```
Comment (trxcllnt) [inline cpp/include/cudf/column/column_device_view.cuh:596]: These are necessary even though this is a `.cuh` file?

I thought `.cuh` files could only be included by `.cu` files, compiled by `nvcc`, so `__CUDACC__` would always be defined?

Comment (codereport) [reply cpp/include/cudf/column/column_device_view.cuh:596]: My understanding is that is the guideline as well. However, this is overwhelming not followed (which I discovered from workin on this PR). For instance, `column_wrapper.hpp` where the `make_counting_transform_iterator` function used to be, has a .hpp, and the majority of the test files that used it were also .hpp.

File: cpp/include/cudf/detail/iterator.cuh [hunk 2]
```diff
@@ -316,5 +320,31 @@ auto inline make_pair_iterator(scalar const& scalar_value)
...
+ * iter[n] == n * n
+ * @endcode
+ *
+ * @param start The starting value of the counting iterator
+ * @param f The unary function to apply to the counting iterator.
+ * This should be a host function and not a device function.
+ * @return auto A transform iterator that applies `f` to a counting iterator
+ */
+template <typename UnaryFunction>
+auto make_counting_transform_iterator(cudf::size_type start, UnaryFunction f)
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/iterator.cuh:344]: `inline` ?

File: cpp/include/cudf/detail/iterator.cuh [hunk 3]
```diff
@@ -37,6 +37,10 @@
 #include <cudf/scalar/scalar.hpp>
 #include <cudf/scalar/scalar_device_view.cuh>
 
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/iterator.cuh:43]: these includes may not be required. previous included files might have already included these.
(that brings us the question, should we include this file's dependencies, even though other files included might already include it?)

File: cpp/tests/binaryop/binop-fixture.hpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
```
Comment (karthikeyann) [inline cpp/tests/binaryop/binop-fixture.hpp:2]: Try my oneliner for updating copyright year in all changes files.
```
gh pr diff | grep -F "+++ b/" | sed "s#+++ b/##g"  | xargs -L1 sed -i -E "s/Copyright *\(c\) *([0-9,-]+)*-([0-9]{4}), *NVIDIA *CORPORATION/Copyright (c) \\1-`date +%Y`, NVIDIA CORPORATION/; /`date +%Y`/! s/Copyright *\(c\) ([0-9]{4}), *NVIDIA *CORPORATION/Copyright (c) \\1-`date +%Y`, NVIDIA CORPORATION/" 
```

File: cpp/tests/strings/booleans_tests.cpp [hunk 5]
```diff
@@ -21,6 +21,8 @@
 #include <cudf_test/column_utilities.hpp>
 #include <cudf_test/column_wrapper.hpp>
 
+#include <thrust/iterator/transform_iterator.h>
+
```
Comment (karthikeyann) [inline cpp/tests/strings/booleans_tests.cpp:25]: is this include needed? there are no other changes in this file.

File: cpp/tests/strings/case_tests.cpp [hunk 6]
```diff
@@ -24,6 +24,8 @@
 #include <cudf_test/column_utilities.hpp>
 #include <cudf_test/column_wrapper.hpp>
 
+#include <thrust/iterator/transform_iterator.h>
+
```
Comment (karthikeyann) [inline cpp/tests/strings/case_tests.cpp:28]: is this include needed? (IDE inserted?)

File: cpp/tests/strings/chars_types_tests.cpp [hunk 7]
```diff
@@ -22,6 +22,8 @@
 #include <cudf_test/column_utilities.hpp>
 #include <cudf_test/column_wrapper.hpp>
 
+#include <thrust/iterator/transform_iterator.h>
+
```
Comment (karthikeyann) [inline cpp/tests/strings/chars_types_tests.cpp:26]: similar includes without any changes are present in many files. check them all at once.

File: cpp/tests/unary/cast_tests.cpp [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2021, NVIDIA CORPORATION.
```
Comment (karthikeyann) [inline cpp/tests/unary/cast_tests.cpp:2]:
```suggestion
 * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/detail/iterator.cuh (hunk 3 above)
Comment (codereport) [reply cpp/include/cudf/detail/iterator.cuh:43]: Yes, ICWYU (Include What You Use). If you you end up modifying the "implicit" dependency, your build breaks.

File: cpp/tests/strings/chars_types_tests.cpp (hunk 7 above)
Comment (codereport) [reply cpp/tests/strings/chars_types_tests.cpp:26]: No I inserted manually.

File: cpp/tests/strings/case_tests.cpp (hunk 6 above)
Comment (codereport) [reply cpp/tests/strings/case_tests.cpp:28]: No I inserted manually.

File: cpp/tests/strings/booleans_tests.cpp (hunk 5 above)
Comment (codereport) [reply cpp/tests/strings/booleans_tests.cpp:25]: No I inserted manually.

File: cpp/include/cudf_test/column_wrapper.hpp [hunk 9]
```diff
@@ -1476,8 +1451,7 @@ class lists_column_wrapper : public detail::column_wrapper {
   void build_from_nested(std::initializer_list<lists_column_wrapper<T, SourceElementT>> elements,
                          std::vector<bool> const& v)
   {
-    auto valids = cudf::test::make_counting_transform_iterator(
-      0, [&v](auto i) { return v.empty() ? true : v[i]; });
+    auto const valids = v.empty() ? std::vector<bool>(elements.size(), true) : v;
```
Comment (karthikeyann) [inline cpp/include/cudf_test/column_wrapper.hpp:1454]: this makes a copy.

**Review by karthikeyann — APPROVED** [review]
LGTM

File: cpp/include/cudf_test/column_wrapper.hpp (hunk 9 above)
Comment (codereport) [reply cpp/include/cudf_test/column_wrapper.hpp]: Reverted

**Comment by codereport** [comment]
rerun tests

File: cpp/include/cudf/column/column_device_view.cuh (hunk 1 above)
Comment (trxcllnt) [reply cpp/include/cudf/column/column_device_view.cuh:596]: So you're saying `column_wrapper.hpp` and other `.hpp` files are doing `#include <cudf/detail/iterator.cuh>`? That seems like a problem.

Comment (codereport) [reply cpp/include/cudf/column/column_device_view.cuh:596]: I agree, but sort of out of scope of this PR. Although ... I just did 
```
cat $(find . | grep "p$") | grep thrust | wc -l
```
in both `cpp/src` and `cpp/include` and the total was 53. However, upon brief review ... most of the references were to `thrust::host_vector` or thrust::make_**_iterator` so maybe those aren't consider "device" code?
___
Just did the same `grep` on `cpp/tests` and got 533 :open_mouth:  
```
cat $(find . | grep "p$") | grep __device__ | wc -l
```
yields 27 across `src` `include` and `tests`.

**Comment by galipremsagar** [comment]
rerun tests

File: cpp/include/cudf/column/column_device_view.cuh (hunk 1 above)
Comment (codereport) [reply cpp/include/cudf/column/column_device_view.cuh:596]: @trxcllnt I will set up a issue / PR to rename filenames / look into this.

**Comment by codereport** [comment]
@gpucibot merge

**Pull request merged**